client/internal/mockserver.go

```diff
@@ -0,0 +1,83 @@
+package internal
+
+import (
+	"log"
+	"net/http"
+	"testing"
+
+	"github.com/gorilla/websocket"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+	"github.com/open-telemetry/opamp-go/internal/testhelpers"
+)
+
+type MockServer struct {
+	Endpoint  string
+	OnRequest func(w http.ResponseWriter, r *http.Request)
+	OnConnect func(r *http.Request, conn *websocket.Conn)
+	OnMessage func(msg *protobufs.AgentToServer) *protobufs.ServerToAgent
+	HTTPSrv   *http.Server
+}
+
+var upgrader = websocket.Upgrader{}
+
+func StartMockServer(t *testing.T) *MockServer {
+	srv := &MockServer{Endpoint: testhelpers.GetAvailableLocalAddress()}
+
+	m := http.NewServeMux()
+	m.HandleFunc(
+		"/", func(w http.ResponseWriter, r *http.Request) {
+			if srv.OnRequest != nil {
+				srv.OnRequest(w, r)
+				return
+			}
+
+			conn, err := upgrader.Upgrade(w, r, nil)
```

Comment:
Why? What does that help with?

Reply:
Sorry, I didn't read this was a test.

The idea I had in mind was to upgrade all connections to a websocket if possible within the middleware instead of doing it per endpoint.

Reply:
It probably does not matter much in this test code, but if you see a room for improvement we can certainly do a follow up PR (a PR from you is also welcome).